### PR TITLE
Ciic 2016 challenge

### LINKED ISSUE #29: CinC/Physionet PCG/ECG challenge 2016
# CinC challenge

https://physionet.org/challenge/2016/ 

A prestigious challenge/conference with nice data! 

:fire: UPDATE: game's still ON! :guitar: 

Looking for hackers to help me set someting up, if it's feasible. The there will be whole summer to tune the app. 

Blocked by: Add encoders #22 
## Plan of attack
- [ ] audio
  - [x] for now use `wav2vect` from Matlab
  - [ ] implement `wavEncoder` - IN PROGRESS #26 
  - [ ] evaluate if functionality of the WAVEncoder (internal scipy) is the same as matlab's
  - [ ] try `Cochlea` encoder
  - [ ] implement sound encoders for nupic.audio #22 
- [ ] training
  - records are Normal/Anomaly/Unknown
  - [ ] aggregate all NORMAL records to a 2 column file (reset, PCG)
    - [x] how radical subsampling? bcs nupic is too slow to process whole dataset: only down to 1000(from 2000),bcs of Sampling Theorem (Fs>=2*F)
  - [ ] commit the training data files (bcs the preprocessing takes long) 
  - [ ] train a HTM model + serialize it
  - [ ] try param swarming
- [ ] evaluation
  - [ ] load the model, disable learning
  - [ ] 2 tasks description.py?, OR other way to train/load/eval a model on datasets
  - [ ] compute average anomaly score for all datapoints of a record
  - [ ] implement the anomaly metric in nupic 
  - [ ] create a model (for nupic?) that does this classification based on avg. anomaly scores?
  - [ ] threshold to Normal/Anomaly/Unknown
- [ ] submission
  - [ ] modify examples `sample2016*`
  - [ ] nupic is installed, so `setup` will just source a virtualenv
  - [ ] each evaluation in `next` will call matlab (wav2csv), python(writes anomaly scores to CSV), matlab again(loads anomalies and decides classification)
  - this is problematic, better go full-python if possible! 
- [ ] improvements:
  - [ ] try bag (multi model) voting
    - [ ] model trained on full normal data
  - [ ] model on FHS parts
  - [ ] model on anomalous data
  - [ ] model pretrained on ECG data from other sources! https://github.com/breznak/nupic.biodat
## Working plan to get some validation results ASAP:
- [ ] training data
  - will train only on `Normal` data and select (FHS) subsequences of it
  - data extracted from Matlab @breznak will do that
- [ ] train HTM model
  - on the provided data  
  - just one HTM model (with RDSE? encoder, what best settings? probably no time to swarm)
  - able to `serialize the model` and `load` to run on eval. data (learning off)
    - the approach with OPF is not reliably working, can someone post code to do that? (@rhyolight or someone..?)
- [ ] write simple classification function: `classify(anScores[])`
  - should decide classification from the anomaly scores for the whole sequence/sample
  - can be sth like `avg` and Normal iff <0.4; UNKNOWN iff [0.4...0.7]; Anomal iff > 0.7; ETA ~10mins
- [ ] score
  - process validation data (@breznak will commit a file)
  - classify & compute score -> submit! :pray:

### PR DESCRIPTION
**Pull request by breznak**
# CinC challenge

https://physionet.org/challenge/2016/ 

A prestigious challenge/conference with nice data! 

:fire: UPDATE: game's still ON! :guitar: 

Looking for hackers to help me set someting up, if it's feasible. The there will be whole summer to tune the app. 

Blocked by: Add encoders #22 
## Plan of attack
- [ ] audio
  - [x] for now use `wav2vect` from Matlab
  - [ ] implement `wavEncoder` - IN PROGRESS #26 
  - [ ] evaluate if functionality of the WAVEncoder (internal scipy) is the same as matlab's
  - [ ] try `Cochlea` encoder
  - [ ] implement sound encoders for nupic.audio #22 
- [ ] training
  - records are Normal/Anomaly/Unknown
  - [ ] aggregate all NORMAL records to a 2 column file (reset, PCG)
    - [x] how radical subsampling? bcs nupic is too slow to process whole dataset: only down to 1000(from 2000),bcs of Sampling Theorem (Fs>=2*F)
  - [ ] commit the training data files (bcs the preprocessing takes long) 
  - [ ] train a HTM model + serialize it
  - [ ] try param swarming
- [ ] evaluation
  - [ ] load the model, disable learning
  - [ ] 2 tasks description.py?, OR other way to train/load/eval a model on datasets
  - [ ] compute average anomaly score for all datapoints of a record
  - [ ] implement the anomaly metric in nupic 
  - [ ] create a model (for nupic?) that does this classification based on avg. anomaly scores?
  - [ ] threshold to Normal/Anomaly/Unknown
- [ ] submission
  - [ ] modify examples `sample2016*`
  - [ ] nupic is installed, so `setup` will just source a virtualenv
  - [ ] each evaluation in `next` will call matlab (wav2csv), python(writes anomaly scores to CSV), matlab again(loads anomalies and decides classification)
  - this is problematic, better go full-python if possible! 
- [ ] improvements:
  - [ ] try bag (multi model) voting
    - [ ] model trained on full normal data
  - [ ] model on FHS parts
  - [ ] model on anomalous data
  - [ ] model pretrained on ECG data from other sources! https://github.com/breznak/nupic.biodat

Fixes #29

### CODE REVIEW

**Comment by breznak** [comment]
@nupic-community/core

**Comment by rhyolight** [comment]
@breznak Looks cool, I wish we would have started this a month ago. 😕

**Comment by breznak** [comment]
@rhyolight yep :/ I got to know about it just "5 mins to deadline". But the actual competition is in the end of the summer, I have to find out..but it would make sense if we could still participate! :pray: I'm going to work on this over the summer, actually I have a working prototype now..

**Comment by rhyolight** [comment]
@breznak Cool! Hey @rcrowder do you mind if we merge this?

File: CinC2016Challenge_PCG/examples/nupic2016/model/description.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+# futures. NOTE: this MUST be called after updateConfigFromSubConfig() in order
+# to support value-getter-based substitutions from the sub-experiment (if any)
+applyValueGettersToContainer(config)
+control = {
+  # The environment that the current model is being run in
+  "environment": 'opfExperiment', #'nupic',
+
+ "tasks":[
+ {
+  "taskLabel": "train",
```
Comment (breznak) [inline CinC2016Challenge_PCG/examples/nupic2016/model/description.py:362]: @rhyolight can you help me with this? Basically I need to train a model, disable learning and run it on evaluation data; 
I'd like to do that with `run_opf_experiment.py` script, I found an example with the `tasks`, but it does not work..not sure if it's supported still, or how to approach?

File: CinC2016Challenge_PCG/examples/nupic2016/main.m [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+enc_resol = mean(abs(diff(pcg_full))) %TODO improve this
+
+%add header
+!echo "reset,consumption" > tr.csv
+!echo "int,float" >> tr.csv
+!echo "R," >> tr.csv
+!cat train.csv >> tr.csv
+
+% run HTM model
+!python run_opf_experiment.py model/ -c trained
```
Comment (breznak) [inline CinC2016Challenge_PCG/examples/nupic2016/main.m:22]: NuPIC does not run like this (paths are set incorrectly when called from matlab), will have to do it the other way: shell script calls python/nupic and matlab code

File: CinC2016Challenge_PCG/examples/nupic2016/main.m [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+!echo "int,float" >> tr.csv
+!echo "R," >> tr.csv
+!cat train.csv >> tr.csv
+
+% run HTM model
+!python run_opf_experiment.py model/ -c trained 
+!python run_opf_experiment.py model/ --load trained --tasks train
+
+%% EVALUATE/CLASSIFY
+!python run_opf_experiment.py model/ --load trained --tasks eval --noCheckpoint
```
Comment (breznak) [inline CinC2016Challenge_PCG/examples/nupic2016/main.m:26]: @rhyolight this is what I was talking about above, I'd like to train model "trained" on task "train", and run model "trained" on task "eval" with learning disabled.

File: CinC2016Challenge_PCG/examples/nupic2016/wav2vect.m [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+%% Load data and resample data
+normals = importfile('normals.csv');
+N=100 %FIXME size(normals)
+prob=0.01; % to make this faster, use only random 2\% of the files; %FIXME process all files
+result=[]; % select signal/features used for training
+signal=[]; % whole PCG signal, unprocessed
+for i=1:N
+    r=normals{i};
+    r=r(2:end-1);
+    [PCG, Fs1] = audioread([r '.wav']);  % load data
```
Comment (breznak) [inline CinC2016Challenge_PCG/examples/nupic2016/wav2vect.m:19]: I'd like to create several audio encoders (#22 ), and be able to use this project as `nupic.vision`, installable through pip. 
- one of the implementations could be based on matlab/octave!

File: CinC2016Challenge_PCG/examples/nupic2016/wav2vect.m [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    if(rand > prob) 
+        continue; %skip
+    end
+    
+    ['At ',num2str(i),' of ',num2str(N),' file: ',r]
+   
+    % Running runSpringerSegmentationAlgorithm.m to obtain the assigned_states
+    [assigned_states] = runSpringerSegmentationAlgorithm(PCG_resampled, springer_options.audio_Fs, Springer_B_matrix, Springer_pi_vector, Springer_total_obs_distribution, false); % obtain the locations for S1, systole, s2 and diastole
+    % use only S1, S2
+    FHS=PCG_resampled(assigned_states==1 | assigned_states==2);
```
Comment (breznak) [inline CinC2016Challenge_PCG/examples/nupic2016/wav2vect.m:35]: ideally, we'd train on full ECG(PCG) signal, but NUPIC is too slow for that (30k+ datapoints per record), have to use FHS and extract only "interesting" parts.

File: CinC2016Challenge_PCG/examples/nupic2016/model/description.py (hunk 1 above)
Comment (rhyolight) [inline CinC2016Challenge_PCG/examples/nupic2016/model/description.py:362]: I'm sorry @breznak but I never use `run_opf_experiment.py`. I found that interface too confusing. Maybe @scottpurdy can help you.

**Comment by rhyolight** [comment]
This is a cool idea, but I don't have any experience using NuPIC with the route you are taking. Let's see if anyone else is interested in helping. @nupic-community/core

File: CinC2016Challenge_PCG/examples/nupic2016/model/description.py (hunk 1 above)
Comment (breznak) [inline CinC2016Challenge_PCG/examples/nupic2016/model/description.py:362]: ...cool, thanks. If not, do you have an example 
- using Regions/Network API and calling run() and save() from code?

**Comment by rcrowder** [comment]
@breznak Looks like a really good start.

**Comment by breznak** [comment]
## Working plan to get some validation results ASAP:
- [ ] training data
  - will train only on `Normal` data and select (FHS) subsequences of it
  - data extracted from Matlab @breznak will do that
- [ ] train HTM model
  - on the provided data  
  - just one HTM model (with RDSE? encoder, what best settings? probably no time to swarm)
  - able to `serialize the model` and `load` to run on eval. data (learning off)
    - the approach with OPF is not reliably working, can someone post code to do that? (@rhyolight or someone..?)
- [ ] write simple classification function: `classify(anScores[])`
  - should decide classification from the anomaly scores for the whole sequence/sample
  - can be sth like `avg` and Normal iff <0.4; UNKNOWN iff [0.4...0.7]; Anomal iff > 0.7; ETA ~10mins
- [ ] score
  - process validation data (@breznak will commit a file)
  - classify & compute score -> submit! :pray: 

CC @fergalbyrne

**Comment by breznak** [comment]
Merging to get collaboration easier
